test/module/shared_model/validators/field_validator_test.cpp

```diff
@@ -0,0 +1,747 @@
+  // Function which initializes field, allows to have one type when dealing
+  // with various types of fields
+  using InitFieldFunction = std::function<void()>;
+
+  /**
+   * FieldTestCase is a struct that represents one value of some field,
+   * and expected validation result.
+   */
+  struct FieldTestCase {
+    std::string name;
+    InitFieldFunction init_func;
+    bool value_is_valid;
+    std::string expected_message;
+  };
+
+  // Returns string containing field name and test case name for debug output
+  std::string testFailMessage(const std::string &field_name,
+                              const std::string &testcase_name) const {
+    return (boost::format("Field: %s\nTest Case: %s\n") % field_name
+            % testcase_name)
+        .str();
+  }
+
+ public:
+  // To test one field, validation function is required,
+  // which is always the same, and several test cases, which represent
+  // various inputs
+  using FieldTest = std::pair<ValidationFunction, std::vector<FieldTestCase>>;
+
+  // Run all test cases for given field
+  void runTestCases(const google::protobuf::FieldDescriptor *field) {
+    auto field_name = field->name();
+    // skip field, if already tested
+    if (checked_fields.find(field_name) != checked_fields.end()) {
+      return;
+    }
+    checked_fields.insert(field_name);
+
+    // Will throw key exception in case new field is added
+    FieldTest field_test;
+    try {
+      field_test = field_validators.at(field_name);
+    } catch (const std::out_of_range &e) {
+      FAIL() << "Missing field setter: " << field_name;
+    }
+    auto validate = field_test.first;
+    for (auto &testcase : field_test.second) {
+      // Initialize field
+      testcase.init_func();
+      // Perform validation
+      auto reason = validate();
+      // if value supposed to be invalid, check that there is a reason
+      // and that error message is as expected.
+      // If value supposed to be valid, check for empty reason.
+      if (!testcase.value_is_valid) {
+        ASSERT_TRUE(!reason.second.empty())
+            << testFailMessage(field_name, testcase.name);
+        EXPECT_EQ(testcase.expected_message, reason.second.at(0))
+            << testFailMessage(field_name, testcase.name);
+      } else {
+        EXPECT_TRUE(reason.second.empty())
+            << testFailMessage(field_name, testcase.name)
+            << "Message: " << reason.second.at(0) << "\n";
+      }
+    }
+  }
+
+ protected:
+  // Because we use Protobuf reflection to generate fields by generating all
+  // possible commands, some fields are checked several times
+  // because they are present in several commands. To prevent that, this set
+  // contains all already tested fields, and checked each time we try to test
+  // a field
+  std::unordered_set<std::string> checked_fields;
+
+  /************************** TEST CASES ***************************/
+
+  std::vector<FieldTestCase> account_id_test_cases{
+      {
```

Comment:
Replacing this with some sort of function which takes case name, account id, bool (which will in turn create correct expected error message) will improve readability and reduce code duplication. Same applies for code below.

Reply:
Do you suggest define a function for each field, which allows test case registration, like "registerAccountIDTestcase(name, account id, should_pass, error_message)"?